File: wpilibjIntegrationTests/src/main/java/edu/wpi/first/wpilibj/command/ConditionalCommandTest.java
package edu.wpi.first.wpilibj.command;

import java.util.logging.Logger;
import static org.junit.Assert.*;

import org.junit.Before;
import org.junit.Test;

import edu.wpi.first.wpilibj.mocks.MockCommand;

public class ConditionalCommandTest extends AbstractCommandTest {
	private static final Logger logger = Logger.getLogger(ConditionalCommandTest.class.getName());
	ConditionalCommand command;
	MockCommand onTrue, onFalse;
	Boolean state;
	@Override
	protected Logger getClassLogger() {
		return logger;
	}
	@Before
	public void initCommands(){
		state = new Boolean(true);
		onTrue = new MockCommand();
		onFalse = new MockCommand();
		command = new ConditionalCommand(() -> this.state, onTrue, onFalse);
	}
	@Test
	public void testOnTrue(){
		state = true;
		Scheduler.getInstance().add(command);
		Scheduler.getInstance().run();//init command and select onTrue
		Scheduler.getInstance().run();//init onTrue
		Scheduler.getInstance().run();
		assertTrue("Did not initialize the true command",onTrue.getInitializeCount()>0);
		assertTrue("Initialized the false command", onFalse.getInitializeCount()==0);
		assertFalse("Conditional Command has not stopped", command.isRunning());
	}
	@Test
	public void testOnFalse(){
		state = false;
		Scheduler.getInstance().add(command);
		Scheduler.getInstance().run();//init command and select onFalse
		Scheduler.getInstance().run();//init onFalse
		Scheduler.getInstance().run();
		assertTrue("Did not initialize the false command", onFalse.getInitializeCount()>0);
		assertTrue("Initialized the true command", onTrue.getInitializeCount()==0);
		assertFalse("Conditional Command has not stopped", command.isRunning());
	}
}

File: wpilibcIntegrationTests/src/command/ConditionalCommandTest.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) FIRST 2014-2016. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include "Commands/Scheduler.h"
#include "Commands/ConditionalCommand.h"
#include "command/MockCommand.h"
#include "gtest/gtest.h"
#include <DriverStation.h>
#include <RobotState.h>

bool condition = true;
bool GetCondition(){
	return condition;
}
class ConditionalCommandTest : public testing::Test {
public:
  ConditionalCommand *command;
  MockCommand *onTrue, *onFalse;
protected:
  virtual void SetUp() override {
	condition = true;
    RobotState::SetImplementation(DriverStation::GetInstance());
    Scheduler::GetInstance()->SetEnabled(true);
    onTrue = new MockCommand();
    onFalse = new MockCommand();
    command = new ConditionalCommand(GetCondition,onTrue, onFalse);
  }

  virtual void TearDown() override {
	  delete(command);
  }
  /**
   * Tears Down the Scheduler at the end of each test.
   * Must be called at the end of each test inside each test in order to prevent
   * them being deallocated
   * when they leave the scope of the test (causing a segfault).
   * This can not be done within the virtual void Teardown() method because it
   * is called outside of the scope of the test.
   */
  void TeardownScheduler() { Scheduler::GetInstance()->ResetAll(); }

};
TEST_F(ConditionalCommandTest, TestOnTrue){
	condition = true;;
	Scheduler::GetInstance() -> AddCommand(command);
	Scheduler::GetInstance() -> Run();
	Scheduler::GetInstance() -> Run();
	Scheduler::GetInstance() -> Run();
	EXPECT_TRUE(onTrue->GetInitializeCount() >0)<<"Did not initialize the true command\n";
	EXPECT_TRUE(onFalse->GetInitializeCount() == 0)<<"Initialized the false command\n";
	EXPECT_FALSE(command->IsRunning())<<"Conditional command has not ended\n";
	TeardownScheduler();
}

TEST_F(ConditionalCommandTest, TestOnFalse){
	condition = false;
	Scheduler::GetInstance() -> AddCommand(command);
	Scheduler::GetInstance() -> Run();
	Scheduler::GetInstance() -> Run();
	Scheduler::GetInstance() -> Run();
	EXPECT_TRUE(onFalse->GetInitializeCount() >0)<<"Did not initialize the false command";
	EXPECT_TRUE(onTrue->GetInitializeCount() == 0)<<"Initialized the true command";
	EXPECT_FALSE(command->IsRunning())<<"Conditional command has not ended";
	TeardownScheduler();
}
